### PR TITLE
Feat/assets

### PR DESCRIPTION
**Pull request by dexturr**
# Assets view

Related to #408 

## Home page changes
Screenshot 2023-09-28 at 17 12 13

## Key details page and accounts
Screenshot 2023-09-22 at 17 28 57
Screenshot 2023-09-22 at 17 29 11

## Key selector
Screenshot 2023-09-22 at 17 28 46

## Empty state
Screenshot 2023-09-28 at 17 12 19

### CODE REVIEW

File: specs/1125-KEYD-key_details.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+## Key details
+
+- The lozenges shown link to that market in Console (<a name="1125-KEYD-001" href="#1125-KEYD-001">1125-KEYD-001</a>)
+- There is a warning to remember that if I hold an open position the balance / totals may not be accurate as is constantly changing (<a name="1125-KEYD-002" href="#1125-KEYD-002">1125-KEYD-002</a>)
+- I can see the balance of each (the sum across ALL account types) (<a name="1125-KEYD-003" href="#1125-KEYD-003">1125-KEYD-003</a>)
+- There is a button / icon that allows me to expand the view to show the breakdown of all non-zero accounts for that asset (<a name="1125-KEYD-004" href="#1125-KEYD-004">1125-KEYD-004</a>)
+- There is a way to switch between keys (or to easily navigate back to the keys page to achieve this) (<a name="1125-KEYD-005" href="#1125-KEYD-005">1125-KEYD-005</a>)
+<!-- Cannot test this as I cannot trigger the menu to open -->
```
Comment (dalebennett1992) [inline specs/1125-KEYD-key_details.md:8]: is this not significant? This is partly a question but also a note to self to have a look during testing

**Comment by dalebennett1992** [comment]
image
So now I am using this, I feel like from a user perspective it would be nice to have a Back button in the top left corner. Coinbase does have this as well in the form of an arrow <- 

You can get back to the main screen by clicking the nav bar at the bottom, but when I use this it feels to me like that it is missing that arrow.

File: frontend/routes/auth/wallets/key-details/accounts-store.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import { RpcMethods } from '../../../../lib/client-rpc-methods'
+import { useAccountsStore } from './accounts-store'
+
+const ACCOUNT_FIXTURE = {
```
Comment (dalebennett1992) [inline frontend/routes/auth/wallets/key-details/accounts-store.spec.tsx:4]: We might end up needing data like this shared at some point but will revisit when we nail down a long term mocking solution for e2e.

File: frontend/routes/auth/wallets/home/key-list.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                onClick={() => onIconClick(k.publicKey)}
+                className="cursor-pointer mt-2 ml-1"
+              >
+                <MessageIcon />
+              </button>
+            }
+          >
+            <NavLink
+              to={{ pathname: `${FULL_ROUTES.wallets}/${k.publicKey}` }}
+              className="hover:bg-vega-dark-200 w-12 h-full border-l border-1 border-vega-dark-150 flex items-center justify-center"
```
Comment (dalebennett1992) [inline frontend/routes/auth/wallets/home/key-list.tsx:57]: I think we should have a locator here if we want this to be nice to automate. I may add, feel free to tweak if one exists when you see this comment

Comment (dalebennett1992) [reply frontend/routes/auth/wallets/home/key-list.tsx:61]: done, feel free to resolve this comment if you're happy enough with what I have put in

File: specs/1125-KEYD-key_details.md (hunk 1 above)
Comment (dexturr) [reply specs/1125-KEYD-key_details.md]: Figured out how to test this now

File: frontend/routes/auth/wallets/key-details/accounts-store.spec.tsx (hunk 2 above)
Comment (dexturr) [reply frontend/routes/auth/wallets/key-details/accounts-store.spec.tsx:4]: Completely agree, but out of scope for this IMO 👍

Comment (dalebennett1992) [reply frontend/routes/auth/wallets/key-details/accounts-store.spec.tsx:4]: oh yeah absolutely

File: test/e2e/key-details.spec.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  beforeEach(async () => {
+    driver = await createWalletAndDriver()
+    apiHelper = new APIHelper(driver)
+    viewWallet = new ViewWallet(driver)
+    keyDetails = new KeyDetails(driver)
+    apiHelper.createKey('Wallet 1', 'Key 2')
+    await navigateToExtensionLandingPage(driver)
+  })
+
+  afterEach(async () => {
```
Comment (dalebennett1992) [inline test/e2e/key-details.spec.ts:23]: I need a PR where I make all of this global teardown.

**Comment by dalebennett1992** [comment]
So I have a couple of issues to consider now that we close the dropdown menu after clicking...

image
1. If I click copy, there is no indication I have copied (it does appear to be copying though)
2. The arrow icon does not take me to block explorer, nor does clicking the key. Renders it being there a bit pointless at present.

I think this poses a similar problem to what you originally mentioned on the view wallets page when the whole wallet was clickable...

I think ideally this works in one of two ways..

1. In the dropdown where we switch keys we remove the option to copy or navigate to block explorer and use it solely to switch between keys
OR...
2. In the dropdown we can click copy and the menu stays open, proceeding to show the tick to indicate we have copied. Links take us to block explorer (I do not know if I think this should stay open or closed)

I am not sure what my preference is but whilst this isn't working as I would expect it makes me wonder if this list tries to do too much, maybe we should just simplify and have the icon and key name and use it as a facility to switch only? Let me know your thoughts.

**Comment by dalebennett1992** [comment]
image
This now works well in terms of switching keys, however once this menu is open I can only get rid of the menu by specifically selecting a key or clicking a navigation tab down the bottom.  So to be clear, I cannot close the menu by clicking the keys dropdown arrow or by clicking off of the menu. I do think this is something we should change before merging as this is a little irritating

**Review by dalebennett1992 — APPROVED** [review]
Approval pending the final change on the back of my feedback and a corresponding unit test plzzz and thx

File: frontend/routes/auth/wallets/key-details/asset-card.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  const { getAssetById } = useAssetsStore((state) => ({
+    getAssetById: state.getAssetById
+  }))
+  const asset = getAssetById(assetId)
+  const { details: { decimals, symbol, name } = {} } = asset
+  if (!decimals || !symbol || !name) throw new Error('Asset details not populated')
+  const filteredAccounts = accounts
+    .filter((a) => !!a.balance && toBigNum(a.balance, +decimals).gt(0))
+    .map((a) => [
+      // TODO how do we handle differences between VegaTypes (from FE mono) and the VegaTypes from REST?
```
Comment (dexturr) [inline frontend/routes/auth/wallets/key-details/asset-card.tsx:79]:
```suggestion
```

**Pull request merged**